pkg/formats/sniffer.go

```diff
@@ -25,72 +34,105 @@ func (fs *Sniffer) SniffReader(f io.ReadSeeker) (Format, error) {
 	fileScanner := bufio.NewScanner(f)
 	fileScanner.Split(bufio.ScanLines)
 
-	formatType := ""
-	formatEncoding := ""
-	formatVersion := ""
+	var format formatDetails
 
 	for fileScanner.Scan() {
-		if strings.Contains(fileScanner.Text(), `"bomFormat"`) && strings.Contains(fileScanner.Text(), `"CycloneDX"`) {
-			formatType = "application/vnd.cyclonedx"
-			formatEncoding = JSON
+		format = fs.sniff(fileScanner.Bytes())
+		if format.Version != "" && format.Type != "" && format.Encoding != "" {
+			break
 		}
+	}
 
-		if strings.Contains(fileScanner.Text(), `"specVersion"`) {
-			parts := strings.Split(fileScanner.Text(), ":")
-			if len(parts) == 2 {
-				ver := strings.TrimPrefix(strings.TrimSuffix(strings.TrimSuffix(strings.TrimSpace(parts[1]), ","), "\""), "\"")
-				if ver != "" {
-					formatVersion = ver
-					formatEncoding = JSON
-				}
-			}
+	fmt.Fprintf(
+		os.Stderr, "format: %s version: %s encoding: %s\n",
+		format.Type, format.Version, format.Encoding,
+	)
+
+	for _, f := range List {
+		if string(f) == fmt.Sprintf("%s+%s;version=%s", format.Type, format.Encoding, format.Version) {
+			return f, nil
 		}
+	}
 
-		if strings.Contains(fileScanner.Text(), "SPDXVersion:") {
-			formatType = "text/spdx"
-			formatEncoding = "text"
+	// TODO(puerco): Implement a light parser in case the string hacks don't work
+	return "", fmt.Errorf("unknown SBOM format")
+}
 
-			for _, ver := range []string{"2.2", "2.3"} {
-				if strings.Contains(fileScanner.Text(), fmt.Sprintf("SPDX-%s", ver)) {
-					formatVersion = ver
-					break
-				}
-			}
-			break
+func (fs *Sniffer) sniff(data []byte) formatDetails {
+	for _, sig := range sniffFormats {
+		if format := sig.sniff(data); format.Type != "" {
+			return format
 		}
+	}
+
+	return formatDetails{}
+}
 
-		// In JSON, the SPDX version field would be quoted
-		if strings.Contains(fileScanner.Text(), "\"spdxVersion\"") ||
-			strings.Contains(fileScanner.Text(), "'spdxVersion'") {
-			formatType = "text/spdx"
-			formatEncoding = JSON
-			if formatVersion != "" {
-				break
+type formatDetails struct {
+	Type     string
+	Version  string
+	Encoding string
+}
```

Comment:
Why introduce another format to capture format info when we have this one? https://github.com/bom-squad/protobom/blob/d56806732603e3d147375f1f2fa33d9d0724ef8d/pkg/formats/formats.go#L10

Reply:
I agree I skipped that,  I was not sure how to set only parts of the format like the sniffer needs. 

Setting the type,version and encoded and seperatly, and I did not want to use 3 repetitive return values along the sniffer..